File: sciurus17_examples_py/sciurus17_examples_py/pick_and_place_tf.py
# Copyright 2026 RT Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

from geometry_msgs.msg import PoseStamped
from moveit.core.robot_state import RobotState
from moveit.planning import MoveItPy
from moveit.planning import PlanRequestParameters
from moveit_msgs.msg import Constraints
from moveit_msgs.msg import JointConstraint
import numpy as np
import rclpy
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node
from scipy.spatial.transform import Rotation
from tf2_ros import TransformException
from tf2_ros import TransformListener
from tf2_ros import TransformStamped
from tf2_ros.buffer import Buffer

from sciurus17_examples_py.utils import plan_and_execute  # noqa: I100


# 左右の腕を識別するための定数クラス
class ArmSide:
    LEFT = 'left'
    RIGHT = 'right'


class PickAndPlaceTf(Node):
    """TFで検出した対象物を掴んで指定位置に置くノード."""

    # グリッパの開閉角度[rad]
    GRIPPER_CLOSE = 0.0
    GRIPPER_OPEN = math.radians(50.0)
    GRIPPER_GRASP = math.radians(20.0)

    # 置く位置（プレース位置）のXYZ[m]
    PLACE_X = 0.35
    PLACE_Y = 0.0
    PLACE_Z = 0.05

    # ピック・プレース動作時のZ軸オフセット[m]
    APPROACH_OFFSET_Z = 0.12  # アプローチ時の高さオフセット
    GRASP_OFFSET_Z = 0.08  # 把持時の高さオフセット

    def __init__(self):
        super().__init__('pick_and_place_tf')
        self.logger = self.get_logger()

        # TFリスナーの初期化
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.tf_past = TransformStamped()

        # MoveItPyのインスタンスを生成し、planning componentを取得
        self.sciurus17 = MoveItPy(node_name='moveit_py')
        self.logger.info('MoveItPy instance created')

        # 首、左右アーム・グリッパ制御用 planning component
        self.neck = self.sciurus17.get_planning_component('neck_group')
        self.l_arm_waist = self.sciurus17.get_planning_component(
            'l_arm_waist_group'
        )
        self.l_gripper = self.sciurus17.get_planning_component(
            'l_gripper_group'
        )
        self.r_arm_waist = self.sciurus17.get_planning_component(
            'r_arm_waist_group'
        )
        self.r_gripper = self.sciurus17.get_planning_component(
            'r_gripper_group'
        )

        # ロボットモデルの取得（ジョイント目標値の設定に使用）
        self.robot_model = self.sciurus17.get_robot_model()

        # プランニングの設定（動作プランナーと速度・加速度スケール）
        self.arm_plan_params = PlanRequestParameters(
            self.sciurus17,
            'ompl_rrtc_default',
        )
        self.gripper_plan_params = PlanRequestParameters(
            self.sciurus17,
            'ompl_rrtc_default',
        )

        # 0.0〜1.0の範囲で設定
        self.arm_plan_params.max_acceleration_scaling_factor = 0.1

        self.arm_plan_params.max_velocity_scaling_factor = 0.1  # 0.0〜1.0の範囲で設定
        # 0.0〜1.0の範囲で設定
        self.gripper_plan_params.max_acceleration_scaling_factor = 1.0

        # 0.0〜1.0の範囲で設定
        self.gripper_plan_params.max_velocity_scaling_factor = 1.0

        # 腰軸の可動範囲を制限する
        self.set_constraints()

        # 初期姿勢に移動
        self.init_body()

        # 0.5秒ごとにon_timerを呼び出すタイマーを作成
        self.timer = self.create_timer(0.5, self.on_timer)

    def on_timer(self):
        # target_0のTF位置姿勢を取得
        try:
            tf_msg = self.tf_buffer.lookup_transform(
                'base_link', 'target_0', rclpy.time.Time()
            )
        except TransformException as ex:
            self.logger.info(f'Could not transform base_link to target: {ex}')
            return

        # TFフィルタリング用の定数
        now = self.get_clock().now()
        FILTERING_TIME = rclpy.duration.Duration(seconds=2)
        STOP_TIME_THRESHOLD = rclpy.duration.Duration(seconds=3)
        DISTANCE_THRESHOLD = 0.01
        TARGET_Z_MIN_LIMIT = 0.02
        TARGET_X_MIN_LIMIT = 0.13
        TARGET_X_MAX_LIMIT = 0.3

        # TF受信からの経過時間と物体の停止時間を計算
        tf_elapsed_time = now - rclpy.time.Time.from_msg(tf_msg.header.stamp)
        tf_stop_time = now - rclpy.time.Time.from_msg(
            self.tf_past.header.stamp
        )

        # 把持対象の位置が可動範囲外の場合は処理しない
        if tf_msg.transform.translation.z < TARGET_Z_MIN_LIMIT:
            return
        if (
            tf_msg.transform.translation.x < TARGET_X_MIN_LIMIT
            or tf_msg.transform.translation.x > TARGET_X_MAX_LIMIT
        ):
            return

        # 現在時刻から2秒以内に受け取ったTFを使用
        if tf_elapsed_time > FILTERING_TIME:
            return

        # 前回のTF位置との差分を計算
        tf_diff = np.linalg.norm(
            [
                self.tf_past.transform.translation.x
                - tf_msg.transform.translation.x,
                self.tf_past.transform.translation.y
                - tf_msg.transform.translation.y,
                self.tf_past.transform.translation.z
                - tf_msg.transform.translation.z,
            ]
        )

        # 把持対象の位置が停止していることを判定
        if tf_diff > DISTANCE_THRESHOLD:
            self.tf_past = tf_msg
            return

        # 把持対象が3秒以上停止している場合ピッキング動作開始
        if tf_stop_time < STOP_TIME_THRESHOLD:
            return

        self.picking(tf_msg.transform.translation)

    def picking(self, target_position):
        # 物体位置に応じて左右の腕を切り替え
        if target_position.y > 0:
            current_arm = ArmSide.LEFT
        else:
            current_arm = ArmSide.RIGHT

        # 何かを掴んでいた時のためにハンドを開閉
        self.move_gripper_angle(current_arm, self.GRIPPER_OPEN)
        self.move_gripper_angle(current_arm, self.GRIPPER_CLOSE)

        # ピック動作（掴みに行く）
        self.control_arm(
            current_arm,
            target_position.x,
            target_position.y,
            target_position.z + self.APPROACH_OFFSET_Z,
        )
        self.move_gripper_angle(current_arm, self.GRIPPER_OPEN)
        self.control_arm(
            current_arm,
            target_position.x,
            target_position.y,
            target_position.z + self.GRASP_OFFSET_Z,
        )
        self.move_gripper_angle(current_arm, self.GRIPPER_GRASP)
        self.control_arm(
            current_arm,
            target_position.x,
            target_position.y,
            target_position.z + self.APPROACH_OFFSET_Z,
        )

        # プレース動作（移動して置く）
        self.control_arm(
            current_arm,
            self.PLACE_X,
            self.PLACE_Y,
            self.PLACE_Z + self.APPROACH_OFFSET_Z,
        )
        self.control_arm(
            current_arm,
            self.PLACE_X,
            self.PLACE_Y,
            self.PLACE_Z + self.GRASP_OFFSET_Z,
        )
        self.move_gripper_angle(current_arm, self.GRIPPER_OPEN)
        self.control_arm(
            current_arm,
            self.PLACE_X,
            self.PLACE_Y,
            self.PLACE_Z + self.APPROACH_OFFSET_Z,
        )

        # 初期姿勢に戻る
        self.init_arm(current_arm)
        self.move_gripper_angle(current_arm, self.GRIPPER_CLOSE)

    def move_gripper_angle(self, current_arm, angle):
        # グリッパを角度[rad]を指定して開閉する
        robot_state = RobotState(self.robot_model)

        if current_arm == ArmSide.LEFT:
            self.l_gripper.set_start_state_to_current_state()
            robot_state.set_joint_group_positions('l_gripper_group', [-angle])
            self.l_gripper.set_goal_state(robot_state=robot_state)
            plan_and_execute(
                self.sciurus17,
                self.l_gripper,
                self.logger,
                single_plan_parameters=self.gripper_plan_params,
            )
        if current_arm == ArmSide.RIGHT:
            self.r_gripper.set_start_state_to_current_state()
            robot_state.set_joint_group_positions('r_gripper_group', [angle])
            self.r_gripper.set_goal_state(robot_state=robot_state)
            plan_and_execute(
                self.sciurus17,
                self.r_gripper,
                self.logger,
                single_plan_parameters=self.gripper_plan_params,
            )

    def control_arm(self, current_arm, x, y, z):
        # アームを目標位置（x, y, z [m]）に動かす
        # 姿勢は左右アームで固定（左腕: roll=-90deg, 右腕: roll=+90deg）
        goal_pose = PoseStamped()
        goal_pose.header.frame_id = 'base_link'
        goal_pose.pose.position.x = x
        goal_pose.pose.position.y = y
        goal_pose.pose.position.z = z

        if current_arm == ArmSide.LEFT:
            quat = Rotation.from_euler(
                'xyz', [-90, 0, 0], degrees=True
            ).as_quat()
            goal_pose.pose.orientation.x = quat[0]
            goal_pose.pose.orientation.y = quat[1]
            goal_pose.pose.orientation.z = quat[2]
            goal_pose.pose.orientation.w = quat[3]
            self.l_arm_waist.set_start_state_to_current_state()
            self.l_arm_waist.set_goal_state(
                pose_stamped_msg=goal_pose, pose_link='l_link7'
            )
            result = plan_and_execute(
                self.sciurus17,
                self.l_arm_waist,
                self.logger,
                single_plan_parameters=self.arm_plan_params,
            )
            return result
        if current_arm == ArmSide.RIGHT:
            quat = Rotation.from_euler(
                'xyz', [90, 0, 0], degrees=True
            ).as_quat()
            goal_pose.pose.orientation.x = quat[0]
            goal_pose.pose.orientation.y = quat[1]
            goal_pose.pose.orientation.z = quat[2]
            goal_pose.pose.orientation.w = quat[3]
            self.r_arm_waist.set_start_state_to_current_state()
            self.r_arm_waist.set_goal_state(
                pose_stamped_msg=goal_pose, pose_link='r_link7'
            )
            result = plan_and_execute(
                self.sciurus17,
                self.r_arm_waist,
                self.logger,
                single_plan_parameters=self.arm_plan_params,
            )
            return result
        return None

    def init_body(self):
        # カメラで把持対象を撮影するための待機姿勢に移動する
        INITIAL_YAW_ANGLE = math.radians(0.0)
        INITIAL_PITCH_ANGLE = math.radians(-80.0)

        # 左腕を初期姿勢に移動
        self.l_arm_waist.set_start_state_to_current_state()
        self.l_arm_waist.set_goal_state(
            configuration_name='l_arm_waist_init_pose'
        )
        plan_and_execute(
            self.sciurus17,
            self.l_arm_waist,
            self.logger,
            single_plan_parameters=self.arm_plan_params,
        )

        # 右腕を初期姿勢に移動
        self.r_arm_waist.set_start_state_to_current_state()
        self.r_arm_waist.set_goal_state(
            configuration_name='r_arm_waist_init_pose'
        )
        plan_and_execute(
            self.sciurus17,
            self.r_arm_waist,
            self.logger,
            single_plan_parameters=self.arm_plan_params,
        )

        # 首を下向きに設定
        joint_values = [
            INITIAL_YAW_ANGLE,
            INITIAL_PITCH_ANGLE,
        ]
        robot_state = RobotState(self.robot_model)
        robot_state.set_joint_group_positions('neck_group', joint_values)
        self.neck.set_start_state_to_current_state()
        self.neck.set_goal_state(robot_state=robot_state)
        plan_and_execute(
            self.sciurus17,
            self.neck,
            self.logger,
            single_plan_parameters=self.arm_plan_params,
        )

    def init_arm(self, current_arm):
        # SRDFに定義された初期姿勢にアームを動かす
        if current_arm == ArmSide.LEFT:
            self.l_arm_waist.set_start_state_to_current_state()
            self.l_arm_waist.set_goal_state(
                configuration_name='l_arm_waist_init_pose'
            )
            plan_and_execute(
                self.sciurus17,
                self.l_arm_waist,
                self.logger,
                single_plan_parameters=self.arm_plan_params,
            )
        if current_arm == ArmSide.RIGHT:
            self.r_arm_waist.set_start_state_to_current_state()
            self.r_arm_waist.set_goal_state(
                configuration_name='r_arm_waist_init_pose'
            )
            plan_and_execute(
                self.sciurus17,
                self.r_arm_waist,
                self.logger,
                single_plan_parameters=self.arm_plan_params,
            )

    def set_constraints(self):
        # 腰軸の可動範囲を制限する
        constraints = Constraints()
        constraints.name = 'arm_constraints'

        joint_constraint = JointConstraint()
        joint_constraint.joint_name = 'waist_yaw_joint'
        joint_constraint.position = 0.0
        joint_constraint.tolerance_above = math.radians(45)
        joint_constraint.tolerance_below = math.radians(45)
        joint_constraint.weight = 1.0
        constraints.joint_constraints.append(joint_constraint)

        self.l_arm_waist.set_path_constraints(path_constraints=constraints)
        self.r_arm_waist.set_path_constraints(path_constraints=constraints)

    def clear_constraints(self):
        # 設定された関節可動制限をクリアする
        self.l_arm_waist.clear_path_constraints()
        self.r_arm_waist.clear_path_constraints()


def main(args=None):
    rclpy.init(args=args)

    pick_and_place_tf_node = PickAndPlaceTf()

    executor = MultiThreadedExecutor()
    executor.add_node(pick_and_place_tf_node)
    executor.spin()

    pick_and_place_tf_node.clear_constraints()

    # Finish with error. Related Issue
    # https://github.com/moveit/moveit2/issues/2693
    pick_and_place_tf_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: sciurus17_examples/src/pick_and_place_tf.cpp
// Copyright 2024 RT Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Reference:
// https://github.com/ros-planning/moveit2_tutorials/blob
// /a547cf49ff7d1fe16a93dfe020c6027bcb035b51/doc/move_group_interface
// /src/move_group_interface_tutorial.cpp
// https://docs.ros.org/en/humble/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Listener-Cpp.html

#include <angles/angles.h>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>

#include <chrono>
#include <cmath>
#include <memory>
#include <vector>

#include <geometry_msgs/msg/pose.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <moveit/move_group_interface/move_group_interface.hpp>
#include <rclcpp/rclcpp.hpp>
#include <tf2/convert.hpp>
#include <tf2/exceptions.hpp>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>

#include "pose_presets.hpp"

using namespace std::chrono_literals;
using MoveGroupInterface = moveit::planning_interface::MoveGroupInterface;

class PickAndPlaceTf : public rclcpp::Node
{
public:
  // 左腕・右腕を区別する列挙型
  enum class ArmSide
  {
    LEFT,
    RIGHT
  };

  // グリッパの開閉角度[rad]
  inline static const double GRIPPER_CLOSE = 0.0;
  inline static const double GRIPPER_OPEN = angles::from_degrees(50.0);
  inline static const double GRIPPER_GRASP = angles::from_degrees(20.0);

  // 置く位置（プレース位置）のXYZ[m]
  inline static const double PLACE_X = 0.35;
  inline static const double PLACE_Y = 0.0;
  inline static const double PLACE_Z = 0.05;

  // 物体位置からの相対オフセット[m]（ピック動作時に使用）
  inline static const double APPROACH_OFFSET_Z = 0.12;
  inline static const double GRASP_OFFSET_Z = 0.08;

  PickAndPlaceTf(
    rclcpp::Node::SharedPtr move_group_neck_node, rclcpp::Node::SharedPtr move_group_l_arm_node,
    rclcpp::Node::SharedPtr move_group_l_gripper_node,
    rclcpp::Node::SharedPtr move_group_r_arm_node,
    rclcpp::Node::SharedPtr move_group_r_gripper_node)
  : Node("pick_and_place_tf_node")
  {
    using namespace std::placeholders;

    // 首のMoveGroupInterfaceを初期化
    move_group_neck_ = std::make_shared<MoveGroupInterface>(move_group_neck_node, "neck_group");
    move_group_neck_->setMaxVelocityScalingFactor(0.1);
    move_group_neck_->setMaxAccelerationScalingFactor(0.1);

    // 左腕のMoveGroupInterfaceを初期化
    move_group_l_arm_ =
      std::make_shared<MoveGroupInterface>(move_group_l_arm_node, "l_arm_waist_group");
    move_group_l_arm_->setMaxVelocityScalingFactor(0.1);
    move_group_l_arm_->setMaxAccelerationScalingFactor(0.1);

    // 左グリッパのMoveGroupInterfaceを初期化
    move_group_l_gripper_ =
      std::make_shared<MoveGroupInterface>(move_group_l_gripper_node, "l_gripper_group");

    // 右腕のMoveGroupInterfaceを初期化
    move_group_r_arm_ =
      std::make_shared<MoveGroupInterface>(move_group_r_arm_node, "r_arm_waist_group");
    move_group_r_arm_->setMaxVelocityScalingFactor(0.1);
    move_group_r_arm_->setMaxAccelerationScalingFactor(0.1);

    // 右グリッパのMoveGroupInterfaceを初期化
    move_group_r_gripper_ =
      std::make_shared<MoveGroupInterface>(move_group_r_gripper_node, "r_gripper_group");

    // SRDFに定義されている姿勢に体全体を初期化
    init_body();

    // 腰軸と首軸の可動範囲を制限
    set_constraints();

    // TFリスナーを初期化（物体位置を取得するため）
    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock(), 2s);
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    // タイマーを起動（500msごとにon_timerを呼び出し）
    timer_ = this->create_wall_timer(500ms, std::bind(&PickAndPlaceTf::on_timer, this));
  }

  // グリッパを角度[rad]を指定して開閉する
  void move_gripper_angle(const ArmSide current_arm, const double angle)
  {
    auto joint_values = move_group_l_gripper_->getCurrentJointValues();

    if (current_arm == ArmSide::LEFT) {
      joint_values[0] = -angle;
      move_group_l_gripper_->setJointValueTarget(joint_values);
      move_group_l_gripper_->move();
    }
    if (current_arm == ArmSide::RIGHT) {
      joint_values[0] = angle;
      move_group_r_gripper_->setJointValueTarget(joint_values);
      move_group_r_gripper_->move();
    }
  }

  // アームを目標位置（x, y, z [m]）に動かす（姿勢は下向き固定）
  void control_arm(const ArmSide current_arm, const double x, const double y, const double z)
  {
    if (current_arm == ArmSide::LEFT) {
      move_group_l_arm_->setPoseTarget(pose_presets::left_arm_downward(x, y, z));
      move_group_l_arm_->move();
    }
    if (current_arm == ArmSide::RIGHT) {
      move_group_r_arm_->setPoseTarget(pose_presets::right_arm_downward(x, y, z));
      move_group_r_arm_->move();
    }
  }

  // SRDFに定義された姿勢名でアームを動かす
  void move_arm_to_named_pose(const ArmSide current_arm, const std::string & name)
  {
    if (current_arm == ArmSide::LEFT) {
      move_group_l_arm_->setNamedTarget(name);
      move_group_l_arm_->move();
    }
    if (current_arm == ArmSide::RIGHT) {
      move_group_r_arm_->setNamedTarget(name);
      move_group_r_arm_->move();
    }
  }

  // アームの関節の一部に可動制限を設定する
  void set_constraints()
  {
    moveit_msgs::msg::Constraints constraints;
    constraints.name = "arm_constraints";

    // 腰軸の可動範囲を±45度に制限する
    moveit_msgs::msg::JointConstraint joint_constraint;
    joint_constraint.joint_name = "waist_yaw_joint";
    joint_constraint.position = 0.0;
    joint_constraint.tolerance_above = angles::from_degrees(45);
    joint_constraint.tolerance_below = angles::from_degrees(45);
    joint_constraint.weight = 1.0;
    constraints.joint_constraints.push_back(joint_constraint);

    move_group_l_arm_->setPathConstraints(constraints);
    move_group_r_arm_->setPathConstraints(constraints);
  }

  // 設定された関節可動制限をクリアする
  void clear_constraints()
  {
    move_group_l_arm_->clearPathConstraints();
    move_group_r_arm_->clearPathConstraints();
  }

private:
  void on_timer()
  {
    // target_0のTF（Transform）を取得して、物体の位置を調べる
    geometry_msgs::msg::TransformStamped tf_msg;

    try {
      // base_linkからtarget_0へのTFを取得
      tf_msg = tf_buffer_->lookupTransform("base_link", "target_0", tf2::TimePointZero);
    } catch (const tf2::TransformException & ex) {
      RCLCPP_INFO(this->get_logger(), "Could not transform base_link to target: %s", ex.what());
      return;
    }

    // 現在時刻を取得
    rclcpp::Time now = this->get_clock()->now();

    // 物体検出のフィルタリング時間（この時間以内のTFのみ有効とする）
    const auto FILTERING_TIME = rclcpp::Duration(2s);
    // 物体が停止していると判定する時間の閾値
    const auto STOP_TIME_THRESHOLD = rclcpp::Duration(3s);
    // 物体が動いているかを判定する距離の閾値[m]
    const double DISTANCE_THRESHOLD = 0.01;
    // 物体のZ座標の最小値[m]（これより低い物体は無視する）
    const double TARGET_Z_MIN_LIMIT = 0.02;
    // 物体のX座標の最小値・最大値[m]（範囲外の物体は無視する）
    const double TARGET_X_MIN_LIMIT = 0.13;
    const double TARGET_X_MAX_LIMIT = 0.3;

    tf2::Stamped<tf2::Transform> tf_current;
    tf2::convert(tf_msg, tf_current);

    // TFのスタンプから経過時間を計算
    const auto tf_elapsed_time = now - rclcpp::Time(tf_msg.header.stamp, RCL_ROS_TIME);
    // 前回のTFから停止している時間を計算
    const auto tf_stop_time =
      now - rclcpp::Time(tf_past_.stamp_.time_since_epoch().count(), RCL_ROS_TIME);

    // 掴む物体の位置がロボットの可動範囲内かをチェック
    if (tf_current.getOrigin().z() < TARGET_Z_MIN_LIMIT) {
      return;
    }
    if (
      tf_current.getOrigin().x() < TARGET_X_MIN_LIMIT ||
      tf_current.getOrigin().x() > TARGET_X_MAX_LIMIT)
    {
      return;
    }

    // 検出されてから2秒以上経過した物体は無視（古い情報のため）
    if (tf_elapsed_time > FILTERING_TIME) {
      return;
    }

    // 前回の位置と現在の位置の距離を計算
    double tf_diff = (tf_past_.getOrigin() - tf_current.getOrigin()).length();

    // 動いている物体は掴まない（距離が閾値より大きい場合）
    if (tf_diff > DISTANCE_THRESHOLD) {
      tf_past_ = tf_current;
      return;
    }

    // 物体が3秒以上停止している場合、ピッキング動作を開始
    if (tf_stop_time < STOP_TIME_THRESHOLD) {
      return;
    }

    picking(tf_current.getOrigin());
  }

  void init_body()
  {
    // 首の初期角度を設定
    const double INITIAL_YAW_ANGLE = angles::from_degrees(0.0);
    const double INITIAL_PITCH_ANGLE = angles::from_degrees(-80.0);

    // 左腕・右腕をSRDFに定義された初期姿勢に移動
    move_group_l_arm_->setNamedTarget("l_arm_waist_init_pose");
    move_group_l_arm_->move();
    move_group_r_arm_->setNamedTarget("r_arm_waist_init_pose");
    move_group_r_arm_->move();

    // 首をやや下向きに設定（物体を見やすくするため）
    std::vector<double> joint_values;
    joint_values.push_back(INITIAL_YAW_ANGLE);
    joint_values.push_back(INITIAL_PITCH_ANGLE);
    move_group_neck_->setJointValueTarget(joint_values);
    move_group_neck_->move();
  }

  void picking(tf2::Vector3 target_position)
  {
    // 物体のY座標に応じて左右の腕を選択
    // Y座標が正の値なら左腕、負の値なら右腕を使用
    ArmSide current_arm;
    if (target_position.y() > 0) {
      current_arm = ArmSide::LEFT;
    } else {
      current_arm = ArmSide::RIGHT;
    }

    // 何かを掴んでいた時のためにグリッパを開閉してリセット
    move_gripper_angle(current_arm, GRIPPER_OPEN);
    move_gripper_angle(current_arm, GRIPPER_CLOSE);

    // ピック動作（物体を掴む）

    // 物体の上方に移動（アプローチ位置）
    control_arm(
      current_arm, target_position.x(), target_position.y(),
      target_position.z() + APPROACH_OFFSET_Z);

    // グリッパを開く
    move_gripper_angle(current_arm, GRIPPER_OPEN);

    // 掴む高さまで下降
    control_arm(
      current_arm, target_position.x(), target_position.y(), target_position.z() + GRASP_OFFSET_Z);

    // グリッパを閉じて物体を掴む
    move_gripper_angle(current_arm, GRIPPER_GRASP);

    // 物体を持ち上げる
    control_arm(
      current_arm, target_position.x(), target_position.y(),
      target_position.z() + APPROACH_OFFSET_Z);

    // プレース動作（物体を置く）

    // プレース位置の上方に移動
    control_arm(current_arm, PLACE_X, PLACE_Y, PLACE_Z + APPROACH_OFFSET_Z);

    // プレース位置まで下降
    control_arm(current_arm, PLACE_X, PLACE_Y, PLACE_Z + GRASP_OFFSET_Z);

    // グリッパを開いて物体を離す
    move_gripper_angle(current_arm, GRIPPER_OPEN);

    // グリッパを少し持ち上げる
    control_arm(current_arm, PLACE_X, PLACE_Y, PLACE_Z + APPROACH_OFFSET_Z);

    // 待機姿勢（初期姿勢）に戻る
    move_arm_to_named_pose(
      current_arm,
      current_arm == ArmSide::LEFT ? "l_arm_waist_init_pose" : "r_arm_waist_init_pose");

    // グリッパを閉じる
    move_gripper_angle(current_arm, GRIPPER_CLOSE);
  }

  // MoveGroupInterfaceのインスタンス（首、左右の腕とグリッパ）
  std::shared_ptr<MoveGroupInterface> move_group_neck_;
  std::shared_ptr<MoveGroupInterface> move_group_l_arm_;
  std::shared_ptr<MoveGroupInterface> move_group_l_gripper_;
  std::shared_ptr<MoveGroupInterface> move_group_r_arm_;
  std::shared_ptr<MoveGroupInterface> move_group_r_gripper_;

  // TFリスナー関連（物体位置を取得するため）
  std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_{nullptr};

  // タイマー（定期的に物体位置をチェックするため）
  rclcpp::TimerBase::SharedPtr timer_{nullptr};

  // 前回取得した物体のTF（位置が停止しているかを判定するため）
  tf2::Stamped<tf2::Transform> tf_past_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions node_options;
  node_options.automatically_declare_parameters_from_overrides(true);

  // 各MoveGroupInterface用のノードを作成
  auto move_group_neck_node = rclcpp::Node::make_shared("move_group_neck_node", node_options);
  auto move_group_l_arm_node = rclcpp::Node::make_shared("move_group_l_arm_node", node_options);
  auto move_group_l_gripper_node =
    rclcpp::Node::make_shared("move_group_l_gripper_node", node_options);
  auto move_group_r_arm_node = rclcpp::Node::make_shared("move_group_r_arm_node", node_options);
  auto move_group_r_gripper_node =
    rclcpp::Node::make_shared("move_group_r_gripper_node", node_options);

  // タイマーとTFリスナーを持つため、MultiThreadedExecutorを使用する
  rclcpp::executors::MultiThreadedExecutor exec;
  auto pick_and_place_tf_node = std::make_shared<PickAndPlaceTf>(
    move_group_neck_node, move_group_l_arm_node, move_group_l_gripper_node, move_group_r_arm_node,
    move_group_r_gripper_node);

  // 各ノードをExecutorに追加
  exec.add_node(pick_and_place_tf_node);
  exec.add_node(move_group_neck_node);
  exec.add_node(move_group_l_arm_node);
  exec.add_node(move_group_l_gripper_node);
  exec.add_node(move_group_r_arm_node);
  exec.add_node(move_group_r_gripper_node);
  exec.spin();

  // 終了時に可動制限をクリア
  pick_and_place_tf_node->clear_constraints();
  rclcpp::shutdown();
  return 0;
}
